src/workers/parser/getOutOfScopeLocations.js

```diff
@@ -38,26 +40,63 @@ function getLocation(func) {
 }
 
 /**
- * Reduces an array of locations to remove items that are completely enclosed
- * by another location in the array.
+ * Find the nearest location containing the input position and
+ * return new locations without inner locations under that nearest location
+ *
+ * @param locations Notice! The locations MUST be sorted by `sortByStart`
+ *                  so that we can do linear time complexity operation.
  */
-function removeOverlaps(
-  locations: AstLocation | AstLocation[],
-  location: AstLocation
-) {
-  // support reducing without an initializing array
-  if (!Array.isArray(locations)) {
-    locations = [locations];
+function removeInnerLocations(locations: AstLocation[], position: AstPosition) {
+  // First, let's find the nearest position-enclosing function location,
+  // which is to loop locations to find the last location enclosing the bp.
+  let nearestPos = findLastIndex(locations, loc =>
+    containsPosition(loc, position)
+  );
+
+  // Second, from the nearest location, loop locations again, stop looping
+  // once seeing the 1st location not enclosed by the nearest location
+  // to find the last inner locations inside the nearest location.
+  let lastInnerPos = -1;
+  if (nearestPos >= 0) {
+    let parentLoc = locations[nearestPos];
+    let firstNotContained = findIndex(
+      locations,
+      loc => !containsLocation(parentLoc, loc),
+      nearestPos + 1
+    );
+    lastInnerPos =
+      firstNotContained < 0 ? locations.length - 1 : firstNotContained - 1;
   }
 
-  const contains =
-    locations.filter(a => containsLocation(a, location)).length > 0;
-
-  if (!contains) {
-    locations.push(location);
+  // Third, remove those inner functions
+  const newLocs = locations.slice();
+  if (nearestPos >= 0 && lastInnerPos >= 0) {
```

Comment:
i dont think it is possible for lastInnerPos to be -1, since you have a check on line 68